src_fun/testing.cpp

```diff
@@ -0,0 +1,483 @@
+                        if (!test.error_wildcard)
+                            Error("Expected a wildcard pattern after 'ERROR WITH'. "
+                                  "If you don't want a pattern, use just 'ERROR' instead");
+                        continue;
+                    }
+                    else if (prefix_equals(line, "ERROR"_s))
+                    {
+                        if (line != "ERROR"_s)
+                            Error("Badly defined test condition, must be either 'SUCCESS', 'ERROR', or 'ERROR WITH <pattern>.'");
+
+                        test.must_error_to_succeed = true;
+                        test.error_wildcard        = "*"_s;
+                        continue;
+                    }
+                    else if (prefix_equals(line, "SUCCESS"_s))
+                    {
+                        if (line != "SUCCESS"_s)
+                            Error("Badly defined test condition, must be either 'SUCCESS', 'ERROR', or 'ERROR WITH <pattern>.'");
+
+                        test.must_error_to_succeed = false;
+                        continue;
+                    }
+                }
+
+                if (!seed_defined)
+                {
+                    Defer(seed_defined = true);
+
+                    if (prefix_equals(line, "SEED"_s))
+                    {
+                        consume_until(&line, "SEED"_s);
+                        test.rng_seed = consume_u64(&line);
+                        if (line)                     Error("Badly defined test seed, seed is defined as SEED <integer>.");
+                        if (test.rng_seed == U64_MAX) Error("Badly defined test seed, seed must be < U64_MAX.");
+
+                        continue;
+                    }
+                    else test.rng_seed = U64_MAX;
+                }
+            }
+
+            started_parsing_description = true;
+            add(&desc_cat, line);
+            add(&desc_cat, "\n"_s);
+        }
+
+        test.description        = resolve_to_string_and_free(&desc_cat, &context->memory);
+        test.first_line_of_code = current_line;
+
+        // parse source until we get to a //# line
+        String_Concatenator code_cat = {};
+        while (true)
+        {
+            add(&code_cat, line_preserved_whitespace);
+            add(&code_cat, "\n"_s);
+            if (!contents || peek_test_line()) break;
+            consume_test_line();
+        }
+
+        test.code = resolve_to_string_and_free(&code_cat, &context->memory);
+
+        if (trim(test.code).length == 0)
+            Error("No code for test ID '%',", test.id);
+
+        add_item(&context->tests, &test);
+        tests_parsed++;
+    }
+
+    if (tests_parsed == 0)
+        Error("No tests in file!");
+
+    return true;
+
+#undef Error
+}
+
+static bool run_code_of_test(Test_Case* test)
+{
+    String_Concatenator code_cat = {};
+    add(&code_cat, "test_assert :: (condition: bool) {} intrinsic \"test_assert\";");
```

Comment:
Instead of crudely prepending the test_assert declaration, you should be adding your declarations to some other preload unit. Then have that be the parent of the test unit.

Then your implementation can comfortably be longer and fully implemented in fun, which removes the unnecessary intrinsic.

This also removes the weird (but currently implicitly satisfied) limitation that code can't start at line 1 without its line numbers getting messed up.